Lack of perfect forwarding triggers compile time check in fmt

It seems that fmt added some extra checks to prevent lifetime errors when passing views as lvalue references. But spdlog doesn't perfect forward arguments, causing unexpected compile time errors when formatting rvalues.

spdlog version: 1.8.0
fmt version: 7.1.0

I'm using the latest version of fmt and speedlog available in vcpkg at this time.

Minimal example:

```
#include <fmt/format.h>
#include <spdlog/spdlog.h>

int main(int argc, char *argv[])
{
	// This works fine
	fmt::print("Arguments: {0}", fmt::join(argv, argv + argc, ", "));

	// I would expect this line to work as well
	spdlog::default_logger()->info("Arguments: {0}", fmt::join(argv, argv + argc, ", "));
}
```

```
cmake_minimum_required(VERSION 3.15)
project(spdlog-error VERSION 1.0.0)

add_executable(spdlog-error "main.cpp")
target_compile_features(spdlog-error INTERFACE cxx_std_20)

find_package(fmt CONFIG REQUIRED)
find_package(spdlog CONFIG REQUIRED)
target_link_libraries(spdlog-error PRIVATE fmt::fmt-header-only
                                           spdlog::spdlog_header_only)
```

Output with MSVC 16.7.7: 
```
[main] Building folder: spdlog-err 
[build] Starting build
[proc] Executing command: D:\Tools\CMake\bin\cmake.EXE --build "d:/src/spdlog-err/build/Visual Studio Community 2019 Release - amd64/Debug" --config Debug --target all --parallel -- -j 10
[build] [1/2  50% :: 2.323] Building CXX object CMakeFiles\spdlog-error.dir\main.cpp.obj
[build] FAILED: CMakeFiles/spdlog-error.dir/main.cpp.obj 
[build] C:\PROGRA~2\MICROS~1\2019\COMMUN~1\VC\Tools\MSVC\1427~1.291\bin\Hostx64\x64\cl.exe  /nologo /TP -DFMT_HEADER_ONLY=1 -DFMT_LOCALE -DFMT_SHARED -DSPDLOG_FMT_EXTERNAL -ID:\Tools\vcpkg\installed\x64-windows\include /DWIN32 /D_WINDOWS /GR /EHsc /Zi /Ob0 /Od /RTC1 -MDd /showIncludes /FoCMakeFiles\spdlog-error.dir\main.cpp.obj /FdCMakeFiles\spdlog-error.dir\ /FS -c ..\..\..\main.cpp
[build] D:\Tools\vcpkg\installed\x64-windows\include\fmt\core.h(1623): error C2338: passing views as lvalues is disallowed
[build] D:\Tools\vcpkg\installed\x64-windows\include\fmt\format.h(3825): note: see reference to function template instantiation 'fmt::v7::format_arg_store<fmt::v7::basic_format_context<fmt::v7::detail::buffer_appender<char>,char>,const fmt::v7::arg_join<It,Sentinel,char>> fmt::v7::make_args_checked<const fmt::v7::arg_join<It,Sentinel,char>&,S,char>(const S &,const fmt::v7::arg_join<It,Sentinel,char> &)' being compiled
[build]         with
[build]         [
[build]             It=char *[],
[build]             Sentinel=char *[],
[build]             S=spdlog::string_view_t
[build]         ]
[build] D:\Tools\vcpkg\installed\x64-windows\include\spdlog/logger.h(340): note: see reference to function template instantiation 'fmt::v7::detail::buffer_appender<char> fmt::v7::format_to<FormatString,const fmt::v7::arg_join<It,Sentinel,char>&,250,char>(fmt::v7::basic_memory_buffer<char,250,std::allocator<char>> &,const S &,const fmt::v7::arg_join<It,Sentinel,char> &)' being compiled
[build]         with
[build]         [
[build]             FormatString=spdlog::string_view_t,
[build]             It=char *[],
[build]             Sentinel=char *[],
[build]             S=spdlog::string_view_t
[build]         ]
[build] D:\Tools\vcpkg\installed\x64-windows\include\spdlog/logger.h(87): note: see reference to function template instantiation 'void spdlog::logger::log_<spdlog::string_view_t,fmt::v7::arg_join<It,Sentinel,char>>(spdlog::source_loc,spdlog::level::level_enum,const FormatString &,const fmt::v7::arg_join<It,Sentinel,char> &)' being compiled
[build]         with
[build]         [
[build]             It=char *[],
[build]             Sentinel=char *[],
[build]             FormatString=spdlog::string_view_t
[build]         ]
[build] D:\Tools\vcpkg\installed\x64-windows\include\spdlog/logger.h(93): note: see reference to function template instantiation 'void spdlog::logger::log<fmt::v7::arg_join<It,Sentinel,char>>(spdlog::source_loc,spdlog::level::level_enum,spdlog::string_view_t,const fmt::v7::arg_join<It,Sentinel,char> &)' being compiled
[build]         with
[build]         [
[build]             It=char *[],
[build]             Sentinel=char *[]
[build]         ]
[build] D:\Tools\vcpkg\installed\x64-windows\include\spdlog/logger.h(111): note: see reference to function template instantiation 'void spdlog::logger::log<char[10],fmt::v7::arg_join<It,Sentinel,char>>(spdlog::level::level_enum,const FormatString (&),const fmt::v7::arg_join<It,Sentinel,char> &)' being compiled
[build]         with
[build]         [
[build]             It=char *[],
[build]             Sentinel=char *[],
[build]             FormatString=char [10]
[build]         ]
[build] ..\..\..\main.cpp(6): note: see reference to function template instantiation 'void spdlog::logger::info<char[10],fmt::v7::arg_join<It,Sentinel,char>>(const FormatString (&),const fmt::v7::arg_join<It,Sentinel,char> &)' being compiled
[build]         with
[build]         [
[build]             It=char *[],
[build]             Sentinel=char *[],
[build]             FormatString=char [10]
[build]         ]
[build] ninja: build stopped: subcommand failed.
[cmakefileapi-parser] Code model version (2.1) of cmake-file-api is unexpected. Expecting (2.0). IntelliSense configuration may be incorrect.
[cmakefileapi-parser] Code model version (2.1) of cmake-file-api is unexpected. Expecting (2.0). IntelliSense configuration may be incorrect.
[build] Build finished with exit code 1
```

Output with gcc 9.3.0 in WSL
```
[1/2] Building CXX object CMakeFiles/spdlog-error.dir/main.cpp.o
FAILED: CMakeFiles/spdlog-error.dir/main.cpp.o
/usr/bin/c++  -DFMT_HEADER_ONLY=1 -DFMT_LOCALE -DSPDLOG_FMT_EXTERNAL -isystem /home/oot/tools/vcpkg/installed/x64-linux/include -O3 -DNDEBUG -MD -MT CMakeFiles/spdlog-error.dir/main.cpp.o -MF CMakeFiles/spdlog-error.dir/main.cpp.o.d -o CMakeFiles/spdlog-error.dir/main.cpp.o -c ../../main.cpp
In file included from /home/oot/tools/vcpkg/installed/x64-linux/include/fmt/format.h:45,
                 from ../../main.cpp:1:
/home/oot/tools/vcpkg/installed/x64-linux/include/fmt/core.h: In instantiation of ‘fmt::v7::format_arg_store<fmt::v7::basic_format_context<fmt::v7::detail::buffer_appender<Char>, Char>, fmt::v7::remove_reference_t<Args>...> fmt::v7::make_args_checked(const S&, fmt::v7::remove_reference_t<Args>& ...) [with Args = {const fmt::v7::arg_join<char**, char**, char>&}; S = fmt::v7::basic_string_view<char>; Char = char]’:
/home/oot/tools/vcpkg/installed/x64-linux/include/fmt/format.h:3825:54:   required from ‘typename fmt::v7::buffer_context<Char>::iterator fmt::v7::format_to(fmt::v7::basic_memory_buffer<Char, SIZE>&, const S&, Args&& ...) [with S = fmt::v7::basic_string_view<char>; Args = {const fmt::v7::arg_join<char**, char**, char>&}; long unsigned int SIZE = 250; Char = char; typename fmt::v7::buffer_context<Char>::iterator = fmt::v7::detail::buffer_appender<char>]’
/home/oot/tools/vcpkg/installed/x64-linux/include/spdlog/logger.h:340:27:   required from ‘void spdlog::logger::log_(spdlog::source_loc, spdlog::level::level_enum, const FormatString&, const Args& ...) [with FormatString = fmt::v7::basic_string_view<char>; Args = {fmt::v7::arg_join<char**, char**, char>}]’
/home/oot/tools/vcpkg/installed/x64-linux/include/spdlog/logger.h:87:9:   required from ‘void spdlog::logger::log(spdlog::source_loc, spdlog::level::level_enum, spdlog::string_view_t, const Args& ...) [with Args = {fmt::v7::arg_join<char**, char**, char>}; spdlog::string_view_t = fmt::v7::basic_string_view<char>]’
/home/oot/tools/vcpkg/installed/x64-linux/include/spdlog/logger.h:93:9:   required from ‘void spdlog::logger::log(spdlog::level::level_enum, const FormatString&, const Args& ...) [with FormatString = char [15]; Args = {fmt::v7::arg_join<char**, char**, char>}]’
/home/oot/tools/vcpkg/installed/x64-linux/include/spdlog/logger.h:111:9:   required from ‘void spdlog::logger::info(const FormatString&, const Args& ...) [with FormatString = char [15]; Args = {fmt::v7::arg_join<char**, char**, char>}]’
../../main.cpp:7:85:   required from here
/home/oot/tools/vcpkg/installed/x64-linux/include/fmt/core.h:1623:53: error: static assertion failed: passing views as lvalues is disallowed
 1621 |       detail::count<(
      |       ~~~~~~~~~~~~~~~
 1622 |               std::is_base_of<detail::view, remove_reference_t<Args>>::value &&
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1623 |               std::is_reference<Args>::value)...>() == 0,
      |               ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~
ninja: build stopped: subcommand failed.
```

Hacking in forwarding allows the program to compile and run as expected.

--- a/include/spdlog/logger.h
+++ b/include/spdlog/logger.h
@@ -75,58 +75,58 @@ public:
 
     // FormatString is a type derived from fmt::compile_string
     template<typename FormatString, typename std::enable_if<fmt::is_compile_string<FormatString>::value, int>::type = 0, typename... Args>
-    void log(source_loc loc, level::level_enum lvl, const FormatString &fmt, const Args &...args)
+    void log(source_loc loc, level::level_enum lvl, const FormatString &fmt, Args&&...args)
     {
-        log_(loc, lvl, fmt, args...);
+        log_(loc, lvl, fmt, std::forward<Args>(args)...);
     }
 
     // FormatString is NOT a type derived from fmt::compile_string but is a string_view_t or can be implicitly converted to one
     template<typename... Args>
-    void log(source_loc loc, level::level_enum lvl, string_view_t fmt, const Args &...args)
+    void log(source_loc loc, level::level_enum lvl, string_view_t fmt, Args&&...args)
     {
-        log_(loc, lvl, fmt, args...);
+        log_(loc, lvl, fmt, std::forward<Args>(args)...);
     }
 
     template<typename FormatString, typename... Args>
-    void log(level::level_enum lvl, const FormatString &fmt, const Args &...args)
+    void log(level::level_enum lvl, const FormatString &fmt, Args&&...args)
     {
-        log(source_loc{}, lvl, fmt, args...);
+        log(source_loc{}, lvl, fmt, std::forward<Args>(args)...);
     }
 
     template<typename FormatString, typename... Args>
-    void trace(const FormatString &fmt, const Args &...args)
+    void trace(const FormatString &fmt, Args&&...args)
     {
-        log(level::trace, fmt, args...);
+        log(level::trace, fmt, std::forward<Args>(args)...);
     }
 
     template<typename FormatString, typename... Args>
-    void debug(const FormatString &fmt, const Args &...args)
+    void debug(const FormatString &fmt, Args&&...args)
     {
-        log(level::debug, fmt, args...);
+        log(level::debug, fmt, std::forward<Args>(args)...);
     }
 
     template<typename FormatString, typename... Args>
-    void info(const FormatString &fmt, const Args &...args)
+    void info(const FormatString &fmt, Args&&...args)
     {
-        log(level::info, fmt, args...);
+        log(level::info, fmt, std::forward<Args>(args)...);
     }
 
     template<typename FormatString, typename... Args>
-    void warn(const FormatString &fmt, const Args &...args)
+    void warn(const FormatString &fmt, Args&&...args)
     {
-        log(level::warn, fmt, args...);
+        log(level::warn, fmt, std::forward<Args>(args)...);
     }
 
     template<typename FormatString, typename... Args>
-    void error(const FormatString &fmt, const Args &...args)
+    void error(const FormatString &fmt, Args&&...args)
     {
-        log(level::err, fmt, args...);
+        log(level::err, fmt, std::forward<Args>(args)...);
     }
 
     template<typename FormatString, typename... Args>
-    void critical(const FormatString &fmt, const Args &...args)
+    void critical(const FormatString &fmt, Args&&...args)
     {
-        log(level::critical, fmt, args...);
+        log(level::critical, fmt, std::forward<Args>(args)...);
     }
 
     template<typename T>
@@ -225,7 +225,7 @@ public:
 #else
 
     template<typename... Args>
-    void log(source_loc loc, level::level_enum lvl, wstring_view_t fmt, const Args &...args)
+    void log(source_loc loc, level::level_enum lvl, wstring_view_t fmt, Args&&...args)
     {
         bool log_enabled = should_log(lvl);
         bool traceback_enabled = tracer_.enabled();
@@ -237,7 +237,7 @@ public:
         {
             // format to wmemory_buffer and convert to utf8
             fmt::wmemory_buffer wbuf;
-            fmt::format_to(wbuf, fmt, args...);
+            fmt::format_to(wbuf, fmt, std::forward<Args>(args)...);
 
             memory_buf_t buf;
             details::os::wstr_to_utf8buf(wstring_view_t(wbuf.data(), wbuf.size()), buf);
@@ -326,7 +326,7 @@ protected:
 
     // common implementation for after templated public api has been resolved
     template<typename FormatString, typename... Args>
-    void log_(source_loc loc, level::level_enum lvl, const FormatString &fmt, const Args &...args)
+    void log_(source_loc loc, level::level_enum lvl, const FormatString &fmt, Args&&...args)
     {
         bool log_enabled = should_log(lvl);
         bool traceback_enabled = tracer_.enabled();
@@ -337,7 +337,7 @@ protected:
         SPDLOG_TRY
         {
             memory_buf_t buf;
-            fmt::format_to(buf, fmt, args...);
+            fmt::format_to(buf, fmt, std::forward<Args>(args)...);
             details::log_msg log_msg(loc, name_, lvl, string_view_t(buf.data(), buf.size()));
             log_it_(log_msg, log_enabled, traceback_enabled);
         }

--- a/include/spdlog/pattern_formatter.h
+++ b/include/spdlog/pattern_formatter.h
@@ -92,9 +92,9 @@ public:
     void format(const details::log_msg &msg, memory_buf_t &dest) override;
 
     template<typename T, typename... Args>
-    pattern_formatter &add_flag(char flag, const Args &...args)
+    pattern_formatter &add_flag(char flag, Args&&...args)
     {
-        custom_handlers_[flag] = details::make_unique<T>(args...);
+        custom_handlers_[flag] = details::make_unique<T>(std::forward<Args>(args)...);
         return *this;
     }
     void set_pattern(std::string pattern);

--- a/include/spdlog/spdlog.h
+++ b/include/spdlog/spdlog.h
@@ -128,51 +128,51 @@ SPDLOG_API spdlog::logger *default_logger_raw();
 SPDLOG_API void set_default_logger(std::shared_ptr<spdlog::logger> default_logger);
 
 template<typename FormatString, typename... Args>
-inline void log(source_loc source, level::level_enum lvl, const FormatString &fmt, const Args &...args)
+inline void log(source_loc source, level::level_enum lvl, const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->log(source, lvl, fmt, args...);
+    default_logger_raw()->log(source, lvl, fmt, std::forward<Args>(args)...);
 }
 
 template<typename FormatString, typename... Args>
-inline void log(level::level_enum lvl, const FormatString &fmt, const Args &...args)
+inline void log(level::level_enum lvl, const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->log(source_loc{}, lvl, fmt, args...);
+    default_logger_raw()->log(source_loc{}, lvl, fmt, std::forward<Args>(args)...);
 }
 
 template<typename FormatString, typename... Args>
-inline void trace(const FormatString &fmt, const Args &...args)
+inline void trace(const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->trace(fmt, args...);
+    default_logger_raw()->trace(fmt, std::forward<Args>(args)...);
 }
 
 template<typename FormatString, typename... Args>
-inline void debug(const FormatString &fmt, const Args &...args)
+inline void debug(const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->debug(fmt, args...);
+    default_logger_raw()->debug(fmt, std::forward<Args>(args)...);
 }
 
 template<typename FormatString, typename... Args>
-inline void info(const FormatString &fmt, const Args &...args)
+inline void info(const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->info(fmt, args...);
+    default_logger_raw()->info(fmt, std::forward<Args>(args)...);
 }
 
 template<typename FormatString, typename... Args>
-inline void warn(const FormatString &fmt, const Args &...args)
+inline void warn(const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->warn(fmt, args...);
+    default_logger_raw()->warn(fmt, std::forward<Args>(args)...);
 }
 
 template<typename FormatString, typename... Args>
-inline void error(const FormatString &fmt, const Args &...args)
+inline void error(const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->error(fmt, args...);
+    default_logger_raw()->error(fmt, std::forward<Args>(args)...);
 }
 
 template<typename FormatString, typename... Args>
-inline void critical(const FormatString &fmt, const Args &...args)
+inline void critical(const FormatString &fmt, Args&&...args)
 {
-    default_logger_raw()->critical(fmt, args...);
+    default_logger_raw()->critical(fmt, std::forward<Args>(args)...);
 }
 
 template<typename T>

